src/crimson/common/gated.h

```diff
@@ -42,14 +53,81 @@ class Gated {
     });
   }
 
+  template <typename Func>
+  auto simple_dispatch(const char* what, Func&& func) {
+    return seastar::with_gate(pending_dispatch, std::forward<Func>(func));
+  }
+
   seastar::future<> close() {
+    assert(seastar::this_shard_id() == sid);
+    gated_logger().debug("closing gate");
     return pending_dispatch.close();
   }
+
   bool is_closed() const {
     return pending_dispatch.is_closed();
   }
+
+  void set_shard_id(seastar::shard_id sid) {
+    this->sid = sid;
+  }
+
+  seastar::shard_id get_shard_id() const {
+    return sid;
+  }
  private:
   seastar::gate pending_dispatch;
+  const seastar::shard_id sid;
+};
+
+// gate_per_shard is a class that provides a gate for each shard.
+// It was introduced to provide a way to have gate for each shard
+// in a seastar application since gates are not supposed to be shared
+// across shards. ( https://tracker.ceph.com/issues/64332 )
+class gate_per_shard {
+ public:
+  gate_per_shard() : gates(seastar::smp::count) {
+    for (unsigned shard = 0; shard < seastar::smp::count; ++shard) {
+      gates[shard].set_shard_id(shard);
+    }
+  }
+  //explicit gate_per_shard(size_t shard_count) : gates(shard_count) {}
+  gate_per_shard(const gate_per_shard&) = delete;
+  gate_per_shard& operator=(const gate_per_shard&) = delete;
+  gate_per_shard(gate_per_shard&&) = default;
+  gate_per_shard& operator=(gate_per_shard&&) = default;
+  ~gate_per_shard() = default;
+
+  template <typename Func, typename T>
+  inline void dispatch_in_background(const char* what, T& who, Func&& func) {
+    (void) dispatch(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func, typename T>
+  inline auto dispatch(const char* what, T& who, Func&& func) {
+    return gates[seastar::this_shard_id()].dispatch(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func>
+  auto simple_dispatch(const char* what, Func&& func) {
+    return gates[seastar::this_shard_id()].simple_dispatch(what, std::forward<Func>(func));
+  }
+
+  bool is_closed() const {
+    return gates[seastar::this_shard_id()].is_closed();
+  }
+
+  seastar::future<> close_all() {
+    ceph_assert(gates.size() == seastar::smp::count);
+    return seastar::parallel_for_each(gates, [] (Gated& gate) {
+      return seastar::smp::submit_to(gate.get_shard_id(), [&gate] {
```

Comment:
Nit, seems `smp::invoke_on_all()` is a better choice in this case. Up to you, though.

Reply:
@NitzanMordhai